src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+
+using Azure;
+using Azure.Core;
+using Azure.Data.AppConfiguration;
+using Microsoft.Extensions.Configuration.AzureAppConfiguration.Constants;
+using Microsoft.Extensions.Configuration.AzureAppConfiguration.Extensions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration.ConfigurationClients
+{
+    internal class FailOverSupportedConfigurationClient : IConfigurationClient
+    {
+        private const int HttpStatusCodeRequestThrottled = 429;
+
+        private readonly IEnumerable<ConfigurationClientState> _clients;
+
+        private readonly TimeSpan _parallelRetryInterval;
+
+        private class ConfigurationClientState
+        {
+            public ConfigurationClientState(Uri endpoint, ConfigurationClient configurationClient)
+            {
+                Endpoint = endpoint;
+                BackoffEndTime = DateTimeOffset.UtcNow;
+                FailedAttempts = 0;
+                Client = configurationClient;
+            }
+
+            public ConfigurationClient Client { get; private set; }
+            public Uri Endpoint { get; private set; }
+            public DateTimeOffset BackoffEndTime { get; set; }
+            public int FailedAttempts { get; set; }
+        }
+
+        public FailOverSupportedConfigurationClient(string connectionString, AzureAppConfigurationOptions options)
+        {
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new ArgumentNullException(nameof(connectionString));
+            }
```

Comment:
Shouldn't need this.

Reply:
This is an internal type. We should have null check everything long before the code can hit here. Do we expect the `connectionString` can ever be null here?

Reply:
> Do we expect the connectionString can ever be null here

No, but can't verify that without looking at all callers, in other types. I'd always expect input validation on public methods even on an internal type. Privates, no.

Reply:
In old times :), we do argument check for public facing methods and do debug.assert for internal methods. This way we don't check the same thing over and over again in a release build, but debug.assert made it clear what's expected in code.

I understand we don't do much debug build nowadays (do we?). I am okay if we take this argument check everywhere practice if that's what we have adopted in other places. We need to be aware of that it adds overheads especially to hot code path.

Reply:
> if that's what we have adopted in other places

Yeah, that's the convention in most of our repos. argument validation for all public methods, debug.asserts for private methods for the reason you mentioned.